token_with_payload/src/contract.rs

```diff
@@ -0,0 +1,290 @@
+            &e,
+            &spender,
+            symbol!("xfer_from"),
+            (&spender_id, nonce, &from, &to, &amount).into_val(&e),
+        );
+        spend_allowance(&e, from.clone(), spender_id, amount.clone());
+        spend_balance(&e, from, amount.clone());
+        receive_balance(&e, to, amount);
+    }
+
+    fn burn(e: Env, admin: Signature, nonce: BigInt, from: Identifier, amount: BigInt) {
+        check_admin(&e, &admin);
+        let admin_id = admin.get_identifier(&e);
+
+        verify_and_consume_nonce(&e, &admin_id, &nonce);
+
+        check_auth(
+            &e,
+            &admin,
+            symbol!("burn"),
+            (admin_id, nonce, &from, &amount).into_val(&e),
+        );
+        spend_balance(&e, from, amount);
+    }
+
+    fn freeze(e: Env, admin: Signature, nonce: BigInt, id: Identifier) {
+        check_admin(&e, &admin);
+        let admin_id = admin.get_identifier(&e);
+
+        verify_and_consume_nonce(&e, &admin_id, &nonce);
+
+        check_auth(
+            &e,
+            &admin,
+            symbol!("freeze"),
+            (admin_id, nonce, &id).into_val(&e),
+        );
+        write_state(&e, id, true);
+    }
+
+    fn mint(e: Env, admin: Signature, nonce: BigInt, to: Identifier, amount: BigInt) {
+        check_admin(&e, &admin);
+        let admin_id = admin.get_identifier(&e);
+
+        verify_and_consume_nonce(&e, &admin_id, &nonce);
+
+        check_auth(
+            &e,
+            &admin,
+            symbol!("mint"),
+            (admin_id, nonce, &to, &amount).into_val(&e),
+        );
+        receive_balance(&e, to, amount);
+    }
+
+    fn set_admin(e: Env, admin: Signature, nonce: BigInt, new_admin: Identifier) {
+        check_admin(&e, &admin);
+        let admin_id = admin.get_identifier(&e);
+
+        verify_and_consume_nonce(&e, &admin_id, &nonce);
+
+        check_auth(
+            &e,
+            &admin,
+            symbol!("set_admin"),
+            (admin_id, nonce, &new_admin).into_val(&e),
+        );
+        write_administrator(&e, new_admin);
+    }
+
+    fn unfreeze(e: Env, admin: Signature, nonce: BigInt, id: Identifier) {
+        check_admin(&e, &admin);
+        let admin_id = admin.get_identifier(&e);
+
+        verify_and_consume_nonce(&e, &admin_id, &nonce);
+
+        check_auth(
+            &e,
+            &admin,
+            symbol!("unfreeze"),
+            (admin_id, nonce, &id).into_val(&e),
+        );
+        write_state(&e, id, false);
+    }
+
+    fn decimals(e: Env) -> u32 {
+        read_decimal(&e)
+    }
+
+    fn name(e: Env) -> Bytes {
+        read_name(&e)
+    }
+
+    fn symbol(e: Env) -> Bytes {
+        read_symbol(&e)
+    }
+}
+
+pub struct TokenPayload;
+
+#[cfg_attr(feature = "export", contractimpl)]
+#[cfg_attr(not(feature = "export"), contractimpl(export = false))]
+impl PayloadTrait for TokenPayload {
+    fn payload(
+        e: Env,
+        function: Symbol,
+        args: Vec<RawVal>,
```

Comment:
For me the main value of the initial proposal for having `fn_payload` was that we have a typed interface to contract fn we need to sign. Here it's still not clear what the args should be, which I think is on of the main issues with signing fns. Or is this complimentary to the initial proposal and contracts still would introduce fn_payload for fns with non-trivial args mapping?

Reply:
There's not really any easy way to do the `f_payload` thing as noted in the PR description. I agree typed interfaces are nice, and I would prefer that. For the sake of this exploration, I went this way because it meant I could move fast without solving other problems (like length limits). The real implementation might look different.